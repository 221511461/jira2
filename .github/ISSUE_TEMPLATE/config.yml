blank_issues_enabled: false
contact_links:
  - name: Support and Bug Requests
    url: https://support.atlassian.com/jira-software-cloud/
    about: Need some help? Please submit a support request or bug report via Jira Software Support portal
