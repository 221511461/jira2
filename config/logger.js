const bunyan = require('bunyan');
const bformat = require('bunyan-format');
// add levelInString to include DEBUG | ERROR | INFO | WARN
const formatOut = bformat({ outputMode: 'bunyan', levelInString: true });

const logger = bunyan.createLogger(
  {
    name: 'github-for-jira',
    stream: formatOut,
    level: 'info',
    serializers: {
      req: require('bunyan-express-serializer'),
    },
  },
);

// Suppress logging in tests
if (process.env.NODE_ENV === 'test') {
  logger.level(bunyan.FATAL + 1);
}

console.debug = logger.debug.bind(logger);
console.error = logger.error.bind(logger);
console.log = logger.info.bind(logger);
console.warn = logger.warn.bind(logger);

module.exports = logger;
