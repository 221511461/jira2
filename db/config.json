{
  "development": {
    "database": "jira-dev",
    "host": "127.0.0.1",
    "dialect": "postgres"
  },
  "test": {
    "database": "jira-test",
    "host": "127.0.0.1",
    "dialect": "postgres"
  },
  "production": {
    "dialect": "postgres",
    "use_env_variable": "PG_DATABASE_URL",
    "disable_sql_logging": true,
    "ssl": false,
    "dialectOptions": {
      "ssl": false
    },
    "pool": {
      "max": 15
    }
  }
}
