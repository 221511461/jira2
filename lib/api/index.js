const express = require('express')
const format = require('date-fns/format')
const app = express()
const bodyParser = require('body-parser').urlencoded({ extended: false })
const octokit = require('@octokit/rest')()

async function getInstallation (client, subscription) {
  const id = subscription.gitHubInstallationId
  try {
    const response = await client.apps.getInstallation({ installation_id: id })
    response.data.syncStatus = subscription.syncStatus
    return response.data
  } catch (err) {
    return { error: err, id, deleted: err.code === 404 }
  }
}

const viewerPermissionQuery = `{
  viewer {
    login
    isEmployee
    organization(login: "github") {
      repository(name: "ecosystem-primitives") {
        viewerPermission
      }
    }
  }
}
`
// All routes require a PAT to belong to someone on staff
// This middleware will take the token and make a request to GraphQL
// to see if it belongs to somone on staff
app.use(async (req, res, next) => {
  const token = req.get('Authorization')
  if (!token) return res.sendStatus(404)
  try {
    // Create a separate octokit instance than the one used by the app
    octokit.authenticate({ type: 'token', token: token.split(' ')[1] })
    const { data, errors } = (await octokit.request({
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json'
      },
      method: 'POST',
      // 'viewer' will be the person that owns the token
      query: viewerPermissionQuery,
      url: '/graphql'
    })).data

    if (errors) {
      res.status(401)
      return res.json({ errors, viewerPermissionQuery })
    }

    if (!data.viewer.organization) {
      console.log(`Non-github scoped token attempted to access staff routes`)
      console.log(`login=${data.viewer.login}, isEmployee=${data.viewer.isEmployee}`)
      res.status(401)
      return res.json({
        error: 'Unauthorized',
        message: 'Token provided does not have access to the `github` organization'
      })
    }

    if (data.viewer.organization.repository.viewerPermission !== 'WRITE') {
      console.log(`User attempted to access staff routes`)
      console.log(`login=${data.viewer.login}, isEmployee=${data.viewer.isEmployee}, viewerPermission=${data.viewer.organization.repository.viewerPermission}`)
      res.status(401)
      return res.json({
        error: 'Unauthorized',
        message: 'Token provided does not have `WRITE` access to the @github/ecosystem-primitives repo.'
      })
    }

    console.log(`Staff routes accessed:`)
    console.log(`login=${data.viewer.login}, isEmployee=${data.viewer.isEmployee}, viewerPermission=${data.viewer.organization.repository.viewerPermission}`)

    next()
  } catch (err) {
    console.log({err})
    if (err.code === 401) {
      res.status(401)
      return res.send(err)
    }
    return res.sendStatus(500)
  }
})

app.get('/', (req, res) => {
  res.send({})
})

app.get('/:installationId', async (req, res) => {
  const { Subscription } = require('../models')
  const { installationId } = req.params
  const { client } = res.locals
  try {
    const subscriptions = await Subscription.getAllForInstallation(installationId)
    if (!subscriptions.length) {
      return res.sendStatus(404)
    }
    const { jiraHost } = subscriptions[0].dataValues
    const installations = await Promise.all(subscriptions.map(subscription => getInstallation(client, subscription)))
    const connections = installations
      .filter(response => !response.error)
      .map(data => ({
        ...data,
        isGlobalInstall: data.repository_selection === 'all',
        updated_at: format(data.updated_at, 'MMMM D, YYYY h:mm a'),
        syncState: data.syncState
      }))

    const failedConnections = installations.filter(response => {
      console.log(response.error)
      return response.error
    })
    res.json({
      host: jiraHost,
      installationId,
      connections,
      failedConnections,
      hasConnections: connections.length > 0 || failedConnections.length > 0,
      repoSyncState: `${req.protocol}://${req.get('host')}/api/${installationId}/repoSyncState.json`
    })
  } catch (err) {
    console.log(err)
    res.status(500)
    return res.json(err)
  }
})

app.get('/:installationId/repoSyncState.json', async (req, res) => {
  const { Subscription } = require('../models')
  try {
    const subscriptions = await Subscription.getAllForInstallation(req.params.installationId)
    if (!subscriptions.length) {
      return res.sendStatus(404)
    }
    const data = subscriptions[0].dataValues.repoSyncState
    return res.json(data)
  } catch (err) {
    res.status(500)
    return res.json(err)
  }
})

app.post('/:installationId/sync', bodyParser, async (req, res) => {
  const { Subscription } = require('../models')
  const { installationId } = req.params
  const { jiraHost } = req.body

  try {
    const subscription = await Subscription.getSingleInstallation(jiraHost, installationId)
    if (!subscription) {
      return res.sendStatus(404)
    }

    await Subscription.findOrStartSync(subscription)

    res.status(202)
    return res.json({
      message: `Successfully (re)started sync for ${installationId}`
    })
  } catch (err) {
    console.log(err)
    return res.sendStatus(401)
  }
})

app.post('/:installationId/migrate/:undo?', bodyParser, async (req, res) => {
  const { Subscription } = require('../models')
  const { installationId } = req.params
  const { jiraHost } = req.body
  const subscription = await Subscription.getSingleInstallation(jiraHost, installationId)

  if (!subscription) {
    return res.sendStatus(404)
  }

  const getJiraClient = require('../jira/client')
  const jiraClient = await getJiraClient(null, installationId, jiraHost)

  try {
    if (req.params.undo) {
      await jiraClient.devinfo.migration.undo()
      await subscription.update({ syncStatus: 'FAILED' })
      res.send(`Successfully called migration undo for ${installationId}`)
      return
    } else {
      await jiraClient.devinfo.migration.complete()
      await subscription.update({ syncStatus: 'COMPLETE' })
      res.send(`Successfully called migration complete for ${installationId}`)
      return
    }
  } catch (err) {
    res.send(`Error trying to complete migration for ${installationId}=${err}`).status(500)
  }
})

module.exports = app
