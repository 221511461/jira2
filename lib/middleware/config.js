const getConfig = require('probot-config')

module.exports = async (context) => {
  context.config = await getConfig(context, 'jira.yml')

  return context
}
