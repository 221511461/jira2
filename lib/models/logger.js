const bunyan = require('bunyan')
const bformat = require('bunyan-format')

module.exports = bunyan.createLogger({
  name: 'sequelize',
  level: 'debug',
  stream: bformat({ outputMode: 'short' })
})
