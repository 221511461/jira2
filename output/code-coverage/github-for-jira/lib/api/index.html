
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for github-for-jira/lib/api</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../index.html">All files</a> github-for-jira/lib/api</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">70.97% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>132/186</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">56% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>28/50</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">57.69% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>15/26</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">73.03% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>130/178</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file medium" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="71.91" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 71%"></div><div class="cover-empty" style="width: 29%"></div></div>
	</td>
	<td data-value="71.91" class="pct medium">71.91%</td>
	<td data-value="178" class="abs medium">128/178</td>
	<td data-value="56" class="pct medium">56%</td>
	<td data-value="50" class="abs medium">28/50</td>
	<td data-value="65.22" class="pct medium">65.22%</td>
	<td data-value="23" class="abs medium">15/23</td>
	<td data-value="73.68" class="pct medium">73.68%</td>
	<td data-value="171" class="abs medium">126/171</td>
	</tr>

<tr>
	<td class="file medium" data-value="serializers.js"><a href="serializers.js.html">serializers.js</a></td>
	<td data-value="50" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 50%"></div><div class="cover-empty" style="width: 50%"></div></div>
	</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="8" class="abs medium">4/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="3" class="abs low">0/3</td>
	<td data-value="57.14" class="pct medium">57.14%</td>
	<td data-value="7" class="abs medium">4/7</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank">istanbul</a>
                at Thu Apr 01 2021 15:06:52 GMT+1100 (Australian Eastern Daylight Time)
            </div>
        </div>
        <script src="../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../sorter.js"></script>
        <script src="../../../block-navigation.js"></script>
    </body>
</html>
    