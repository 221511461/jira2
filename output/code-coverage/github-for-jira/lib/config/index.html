
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for github-for-jira/lib/config</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../index.html">All files</a> github-for-jira/lib/config</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">98.67% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>74/75</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">59.09% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>13/22</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">100% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>16/16</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">98.61% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>71/72</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
    </div>
    <div class='status-line high'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="cidr-validator.js"><a href="cidr-validator.js.html">cidr-validator.js</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="11" class="abs high">11/11</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="10" class="abs high">10/10</td>
	</tr>

<tr>
	<td class="file high" data-value="enhance-octokit.js"><a href="enhance-octokit.js.html">enhance-octokit.js</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="22" class="abs high">22/22</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="2" class="abs medium">1/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="22" class="abs high">22/22</td>
	</tr>

<tr>
	<td class="file high" data-value="github-api.js"><a href="github-api.js.html">github-api.js</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="14" class="abs high">14/14</td>
	<td data-value="71.43" class="pct medium">71.43%</td>
	<td data-value="7" class="abs medium">5/7</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="14" class="abs high">14/14</td>
	</tr>

<tr>
	<td class="file high" data-value="redis-info.js"><a href="redis-info.js.html">redis-info.js</a></td>
	<td data-value="87.5" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 87%"></div><div class="cover-empty" style="width: 13%"></div></div>
	</td>
	<td data-value="87.5" class="pct high">87.5%</td>
	<td data-value="8" class="abs high">7/8</td>
	<td data-value="60" class="pct medium">60%</td>
	<td data-value="10" class="abs medium">6/10</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="85.71" class="pct high">85.71%</td>
	<td data-value="7" class="abs high">6/7</td>
	</tr>

<tr>
	<td class="file high" data-value="statsd.js"><a href="statsd.js.html">statsd.js</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="20" class="abs high">20/20</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="2" class="abs medium">1/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="5" class="abs high">5/5</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="19" class="abs high">19/19</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank">istanbul</a>
                at Thu Apr 01 2021 15:06:52 GMT+1100 (Australian Eastern Daylight Time)
            </div>
        </div>
        <script src="../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../sorter.js"></script>
        <script src="../../../block-navigation.js"></script>
    </body>
</html>
    