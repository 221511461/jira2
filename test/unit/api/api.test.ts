/* eslint-disable @typescript-eslint/no-explicit-any */
import supertest from "supertest";
import express, { Application, NextFunction, Request, Response } from "express";
import { Installation, RepoSyncState, Subscription } from "../../../src/models";
import InstallationClass from "../../../src/models/installation";
import SubscriptionClass from "../../../src/models/subscription";
import api from "../../../src/api";
import { getLogger } from "../../../src/config/logger";
import getAxiosInstance from "../../../src/jira/client/axios";
import { mocked } from "ts-jest/utils";

jest.mock("../../../src/config/feature-flags");
jest.mock("../../../src/jira/client/axios");

describe("API", () => {
	let app: Application;
	let locals;
	const invalidId = 99999999;
	const gitHubInstallationId = 1234;
	let installation: InstallationClass;
	let subscription: SubscriptionClass;

	const successfulAuthResponseWrite = {
		data: {
			viewer: {
				login: "gimenete",
				organization: {
					viewerCanAdminister: true
				}
			}
		}
	};

	const successfulAuthResponseAdmin = {
		data: {
			viewer: {
				login: "monalisa",
				organization: {
					viewerCanAdminister: true
				}
			}
		}
	};

	const createApp = async () => {
		const app = express();
		app.use((req: Request, res: Response, next: NextFunction) => {
			res.locals = locals || {};
			req.log = getLogger("test");
			req.session = { jiraHost };
			next();
		});
		app.use("/api", api);
		return app;
	};

	beforeEach(async () => {
		locals = {
			client: {
				apps: {
					getInstallation: jest.fn().mockResolvedValue({ data: {} })
				}
			}
		};
		app = await createApp();

		installation = await Installation.create({
			gitHubInstallationId,
			jiraHost,
			sharedSecret: "secret",
			clientKey: "client-key"
		});

		subscription = await Subscription.create({
			gitHubInstallationId,
			jiraHost,
			jiraClientKey: "client-key",
			repoSyncState: {
				jiraHost,
				numberOfSyncedRepos: 1,
				repos: {
					"1": {
						pullStatus: "complete",
						branchStatus: "complete",
						commitStatus: "complete",
						lastBranchCursor: "foo",
						lastCommitCursor: "bar",
						lastPullCursor: 12,
						repository: {
							id: "1",
							name: "github-for-jira",
							full_name: "atlassian/github-for-jira",
							html_url: "github.com/atlassian/github-for-jira",
							owner: {
								login: "atlassian"
							},
							updated_at: new Date(0)
						}
					}
				}
			}
		});
	});

	afterEach(async () => {
		await Installation.destroy({ truncate: true });
		await Subscription.destroy({ truncate: true });
	});

	describe("Authentication", () => {
		it("should return 404 if no token is provided", () => {
			return supertest(app)
				.get("/api")
				.expect(404)
				.then((response) => {
					expect(response.body).toMatchSnapshot();
				});
		});

		it("should return 200 if a valid token is provided", () => {
			githubNock
				.post("/graphql")
				.reply(200, successfulAuthResponseWrite);

			return supertest(app)
				.get("/api")
				.set("Authorization", "Bearer xxx")
				.expect(200)
				.then((response) => {
					expect(response.body).toMatchSnapshot();
				});
		});

		it("should return 200 if token belongs to an admin", () => {
			githubNock
				.post("/graphql")
				.reply(200, successfulAuthResponseAdmin);

			return supertest(app)
				.get("/api")
				.set("Authorization", "Bearer xxx")
				.expect(200)
				.then((response) => {
					expect(response.body).toMatchSnapshot();
				});
		});

		it("should return 401 if the GraphQL query returns errors", () => {
			githubNock
				.post("/graphql")
				.reply(200, {
					errors: [
						{
							path: ["query", "viewer", "foo"],
							extensions: {
								code: "undefinedField",
								typeName: "User",
								fieldName: "foo"
							},
							locations: [
								{
									line: 4,
									column: 5
								}
							],
							message: "Field 'foo' doesn't exist on type 'User'"
						}
					]
				});

			return supertest(app)
				.get("/api")
				.set("Authorization", "Bearer xxx")
				.then((response) => {
					expect(response.body).toMatchSnapshot();
					expect(response.status).toEqual(401);
				});
		});

		it("should return 401 if the returned organization is null", () => {
			githubNock
				.post("/graphql")
				.reply(200, {
					data: {
						viewer: {
							login: "gimenete",
							organization: null
						}
					}
				});

			return supertest(app)
				.get("/api")
				.set("Authorization", "Bearer xxx")
				.expect(401)
				.then((response) => {
					expect(response.body).toMatchSnapshot();
				});
		});

		it("should return 401 if the token is invalid", () => {
			githubNock
				.post("/graphql")
				.reply(401, {
					HttpError: {
						message: "Bad credentials",
						documentation_url: "https://developer.github.com/v4"
					}
				});

			return supertest(app)
				.get("/api")
				.set("Authorization", "Bearer bad token")
				.expect(401)
				.then((response) => {
					expect(response.body).toMatchSnapshot();
				});
		});
	});

	describe("Endpoints", () => {
		function mockJiraResponse(status: number) {
			// eslint-disable-next-line @typescript-eslint/ban-ts-comment
			// @ts-ignore
			mocked(getAxiosInstance).mockReturnValue({
				"get": () => Promise.resolve<any>({
					status
				})
			});
		}

		describe("verify", () => {
			beforeEach(async () => {
				mockJiraResponse(200);
			});

			it("should return 'Installation already enabled'", () => {
				return supertest(app)
					.post(`/api/jira/${installation.id}/verify`)
					.set("Authorization", "Bearer xxx")
					.expect(200)
					.expect("Content-Type", /json/)
					.then((response) => {
						expect(response.body.message).toMatchSnapshot();
					});
			});
		});

		beforeEach(() => {
			githubNock
				.post("/graphql")
				.reply(200, successfulAuthResponseWrite);
		});

		describe("installation", () => {
			it("should return 404 if no installation is found", async () => {
				return supertest(app)
					.get(`/api/${invalidId}`)
					.set("Authorization", "Bearer xxx")
					.expect(404)
					.then((response) => {
						expect(response.body).toMatchSnapshot();
					});
			});

			it("should return information for an existing installation", async () => {
				return supertest(app)
					.get(`/api/${gitHubInstallationId}`)
					.set("Authorization", "Bearer xxx")
					.set("host", "127.0.0.1")
					.send(`jiraHost=${jiraHost}`)
					.expect(200)
					.then((response) => {
						expect(response.body).toMatchSnapshot();
					});
			});
		});

		describe("repoSyncState", () => {
			beforeEach(async () => {
				await RepoSyncState.create({
					subscriptionId: subscription.id,
					repoId: 1,
					repoName: "github-for-jira",
					repoOwner: "atlassian",
					repoFullName: "atlassian/github-for-jira",
					repoUrl: "github.com/atlassian/github-for-jira",
					branchStatus: "complete",
					branchCursor: "foo",
					commitStatus: "complete",
					commitCursor: "bar",
					pullStatus: "complete",
					pullCursor: "12",
					repoUpdatedAt: new Date(0)
				});
			});

			afterEach(async () => {
				await RepoSyncState.destroy({ truncate: true });
			});

			it("should return 404 if no installation is found", async () => {
				return supertest(app)
					.get(`/api/${invalidId}/repoSyncState.json`)
					.set("Authorization", "Bearer xxx")
					.set("host", "127.0.0.1")
					.send(`jiraHost=${jiraHost}`)
					.expect(404)
					.then((response) => {
						expect(response.body).toMatchSnapshot();
					});
			});

			it("should return the repoSyncState information for an existing installation", async () => {
				return supertest(app)
					.get(`/api/${gitHubInstallationId}/${encodeURIComponent(jiraHost)}/repoSyncState.json`)
					.set("Authorization", "Bearer xxx")
					.set("host", "127.0.0.1")
					.expect(200)
					.then((response) => {
						expect(response.body).toMatchObject({
							jiraHost,
							numberOfSyncedRepos: 1,
							repos: {
								"1": {
									pullStatus: "complete",
									branchStatus: "complete",
									commitStatus: "complete",
									lastBranchCursor: "foo",
									lastCommitCursor: "bar",
									lastPullCursor: 12,
									repository: {
										id: "1",
										name: "github-for-jira",
										full_name: "atlassian/github-for-jira",
										html_url: "github.com/atlassian/github-for-jira",
										owner: {
											login: "atlassian"
										},
										updated_at: new Date(0).toISOString()
									}
								}
							}
						});
					});
			});
		});

		describe("sync", () => {
			it("should return 404 if no installation is found", async () => {
				return supertest(app)
					.post(`/api/${invalidId}/sync`)
					.set("Authorization", "Bearer xxx")
					.send("jiraHost=https://unknownhost.atlassian.net")
					.expect(404)
					.then((response) => {
						expect(response.text).toMatchSnapshot();
					});
			});

			it("should trigger the sync or start function", async () => {
				return supertest(app)
					.post(`/api/${gitHubInstallationId}/sync`)
					.set("Authorization", "Bearer xxx")
					.set("host", "127.0.0.1")
					.send(`jiraHost=${jiraHost}`)
					.expect(202)
					.then((response) => {
						expect(response.text).toMatchSnapshot();
					});
			});

			it("should reset repoSyncState if asked to", async () => {
				return supertest(app)
					.post(`/api/${gitHubInstallationId}/sync`)
					.set("Authorization", "Bearer xxx")
					.set("host", "127.0.0.1")
					.send(`jiraHost=${jiraHost}`)
					.send("resetType=full")
					.expect(202)
					.then((response) => {
						expect(response.text).toMatchSnapshot();
						// td.verify(Subscription.findOrStartSync(subscription, "full"));
					});
			});
		});
	});
});
