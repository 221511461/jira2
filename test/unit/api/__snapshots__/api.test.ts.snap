// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API Authentication should return 200 if a valid token is provided 1`] = `Object {}`;

exports[`API Authentication should return 200 if token belongs to an admin 1`] = `Object {}`;

exports[`API Authentication should return 401 if the GraphQL query returns errors 1`] = `
Object {
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "undefinedField",
        "fieldName": "foo",
        "typeName": "User",
      },
      "locations": Array [
        Object {
          "column": 5,
          "line": 4,
        },
      ],
      "message": "Field 'foo' doesn't exist on type 'User'",
      "path": Array [
        "query",
        "viewer",
        "foo",
      ],
    },
  ],
  "viewerPermissionQuery": "{
  viewer {
    login
    organization(login: \\"fusion-arc\\") {
      viewerCanAdminister
    }
  }
}
",
}
`;

exports[`API Authentication should return 401 if the returned organization is null 1`] = `
Object {
  "error": "Unauthorized",
  "message": "Token provided does not have required access",
}
`;

exports[`API Authentication should return 401 if the token is invalid 1`] = `
Object {
  "documentation_url": "https://developer.github.com/v4",
  "message": "Bad credentials",
}
`;

exports[`API Authentication should return 404 if no token is provided 1`] = `Object {}`;

exports[`API Endpoints installation should return 404 if no installation is found 1`] = `Object {}`;

exports[`API Endpoints installation should return information for an existing installation 1`] = `
Object {
  "connections": Array [
    Object {
      "isGlobalInstall": false,
      "syncStatus": null,
      "updated_at": "Invalid Date",
    },
  ],
  "failedConnections": Array [],
  "hasConnections": true,
  "host": "https://test-atlassian-instance.net",
  "installationId": "1234",
  "repoSyncState": "http://127.0.0.1/api/1234/repoSyncState.json",
}
`;

exports[`API Endpoints repoSyncState should return 404 if no installation is found 1`] = `Object {}`;

exports[`API Endpoints repoSyncState should return the repoSyncState information for an existing installation 1`] = `
Object {
  "jiraHost": "https://test-atlassian-instance.net",
  "numberOfSyncedRepos": 1,
  "repos": Object {
    "1": Object {
      "branchStatus": "complete",
      "commitStatus": "complete",
      "lastBranchCursor": "foo",
      "lastCommitCursor": "bar",
      "lastPullCursor": 12,
      "pullStatus": "complete",
      "repository": Object {
        "full_name": "atlassian/github-for-jira",
        "html_url": "github.com/atlassian/github-for-jira",
        "id": "1",
        "name": "github-for-jira",
        "owner": Object {
          "login": "atlassian",
        },
        "updated_at": "1970-01-01T00:00:00.000Z",
      },
    },
  },
}
`;

exports[`API Endpoints sync should reset repoSyncState if asked to 1`] = `"{\\"message\\":\\"Successfully (re)started sync for 1234\\"}"`;

exports[`API Endpoints sync should return 404 if no installation is found 1`] = `"Not Found"`;

exports[`API Endpoints sync should trigger the sync or start function 1`] = `"{\\"message\\":\\"Successfully (re)started sync for 1234\\"}"`;

exports[`API Endpoints verify should return 'Installation already enabled' 1`] = `"Verification successful"`;
